subsys/mgmt/mcumgr/serial_util.c

```diff
@@ -157,179 +156,158 @@ static int mcumgr_serial_tx_small(const void *data, int len,
 {
+	bool first = true;
+	bool last = false;
 	uint8_t raw[3];
 	uint16_t u16;
+	uint16_t crc;
 	int src_off = 0;
 	int rc = 0;
 	int to_process;
-	bool last = false;
 	int reminder;
 
 	/*
 	 * This is max input bytes that can be taken to the frame before encoding with Base64;
 	 * Base64 has three-to-four ratio, which means that for each three input bytes there are
-	 * going to be four bytes of output used.  The frame starts with two byte marker and ends
+	 * going to be four bytes of output used. The frame starts with two byte marker and ends
 	 * with newline character, which are not encoded (this is the "-3" in calculation).
 	 */
 	int max_input = (((MCUMGR_SERIAL_MAX_FRAME - 3) >> 2) * 3);
 
-	if (first) {
-		/* First frame marker */
-		u16 = sys_cpu_to_be16(MCUMGR_SERIAL_HDR_PKT);
-	} else {
-		/* Continuation frame marker */
-		u16 = sys_cpu_to_be16(MCUMGR_SERIAL_HDR_FRAG);
-	}
-
-	rc = cb(&u16, sizeof(u16), arg);
-	if (rc != 0) {
-		return rc;
-	}
+	/* Calculate the CRC16 checksum of the whole packet, prior to splitting it into frames */
+	crc = mcumgr_serial_calc_crc(data, len);
 
-	/*
-	 * Only the first fragment contains the packet length; the packet length, which is two
-	 * byte long is paired with first byte of input buffer to form triplet for Base64 encoding.
-	 */
-	if (first) {
-		/* The size of the CRC16 should be added to packet length */
-		u16 = sys_cpu_to_be16(len + 2);
-		memcpy(raw, &u16, sizeof(u16));
-		raw[2] = data[0];
+	/* First frame marker */
+	u16 = sys_cpu_to_be16(MCUMGR_SERIAL_HDR_PKT);
 
-		rc = mcumgr_serial_tx_small(raw, 3, cb, arg);
+	while (src_off < len) {
+		/* Send first frame or continuation frame marker */
+		rc = cb(&u16, sizeof(u16));
 		if (rc != 0) {
 			return rc;
 		}
 
-		++src_off;
-		/* One triple of allowed input already used */
-		max_input -= 3;
-	}
+		/*
+		 * Only the first fragment contains the packet length; the packet length, which is
+		 * two byte long is paired with first byte of input buffer to form triplet for
+		 * Base64 encoding.
+		 */
+		if (first) {
+			/* The size of the CRC16 should be added to packet length */
+			u16 = sys_cpu_to_be16(len + 2);
+			memcpy(raw, &u16, sizeof(u16));
+			raw[2] = data[0];
 
-	/*
-	 * Only as much as fits into the frame can be processed, but we also need to fit in the
-	 * two byte CRC.  The frame can not be stretched and current logic does not allow to
-	 * send CRC separately so if CRC would not fit as a whole, shrink to_process by one byte
-	 * forcing one byte to the next frame, with the CRC.
-	 */
-	to_process = MIN(max_input, len - src_off);
-	reminder = max_input - (len - src_off);
+			rc = mcumgr_serial_tx_small(raw, 3, cb);
+			if (rc != 0) {
+				return rc;
+			}
 
-	if (reminder == 0 || reminder == 1) {
-		to_process -= 1;
-	} else if (reminder >= 2) {
-		last = true;
-	}
+			++src_off;
+			/* One triple of allowed input already used */
+			max_input -= 3;
+		}
```

Comment:
Unwind the loop a little bit and take the "first" case out of it. Save us some conditional checking within loop.

Reply:
This would still need a check if it was removed as taking it out would need the start or continuation frame marking sending, and then an if would need to be added to the loop to see if it was sent (first packet) or not